Client/Frontend/Browser/SearchLoader.swift

```diff
@@ -0,0 +1,81 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+import Foundation
+import Shared
+import Storage
+
+private let URLBeforePathRegex = NSRegularExpression(pattern: "^https?://([^/]+/)", options: nil, error: nil)!
+
+// TODO: Swift current requires that classes extending generic classes must also be generic.
+// This is a workaround until that requirement is fixed.
+typealias SearchLoader = _SearchLoader<AnyObject, AnyObject>
+
+/**
+ * Shared data source for the SearchViewController and the URLBar domain completion.
+ * Since both of these use the same SQL query, we can perform the query once and dispatch the results.
+ */
+class _SearchLoader<UnusedA, UnusedB>: Loader<Cursor, SearchViewController> {
+    private let history: History
+    private let urlBar: URLBarView
+
+    init(history: History, urlBar: URLBarView) {
+        self.history = history
+        self.urlBar = urlBar
+        super.init()
+    }
+
+    var query: String = "" {
+        didSet {
+            if query.isEmpty {
+                self.load(Cursor(status: .Success, msg: "Empty query"))
+                return
+            }
+
+            let options = QueryOptions(filter: query, filterType: FilterType.Url, sort: QuerySort.Frecency)
+            self.history.get(options, complete: { (cursor: Cursor) in
+                if cursor.status != .Success {
+                    println("Err: \(cursor.statusMessage)")
+                } else {
+                    self.load(cursor)
+                    self.urlBar.setAutocompleteSuggestion(self.getAutocompleteSuggestion(cursor))
+                }
+            })
+        }
+    }
+
+    private func getAutocompleteSuggestion(cursor: Cursor) -> String? {
```

Comment:
And strange that this Loader holds Autocomplete stuff in it. Why can't the urlbar be a loaderListener?... looking at your comments, but I'm not quite sure if they mention this.